common/matrix/batch_csr_kernels.hpp.inc

```diff
@@ -42,21 +42,51 @@ __device__ __forceinline__ void spmv_kernel(
     const gko::batch_dense::BatchEntry<const ValueType> &b,
     const gko::batch_dense::BatchEntry<ValueType> &c)
 {
-    for (int row = threadIdx.x; row < a.num_rows; row += blockDim.x) {
-        for (int j = 0; j < c.num_rhs; ++j) {
-            c.values[row * c.stride + j] = zero<ValueType>();
+    constexpr auto tile_size = config::warp_size;
+
+    auto thread_block = group::this_thread_block();
+
+    auto subwarp_grp = group::tiled_partition<tile_size>(thread_block);
+
+    const int subwarp_grp_id = floor(static_cast<double>(threadIdx.x) /
+                                     static_cast<double>(tile_size));
+
+    const int num_subwarp_grps_per_block =
+        ceil(static_cast<double>(blockDim.x) / static_cast<double>(tile_size));
+
+
+    for (int row_and_rhs_combination = subwarp_grp_id;
+         row_and_rhs_combination < a.num_rows * b.num_rhs;
+         row_and_rhs_combination += num_subwarp_grps_per_block) {
+        int row = row_and_rhs_combination / b.num_rhs;
+        int rhs = row_and_rhs_combination % b.num_rhs;
+
+        int row_start = a.row_ptrs[row];
+        int row_end = a.row_ptrs[row + 1];
+
+        ValueType temp = zero<ValueType>();
+
+
+        for (int i = subwarp_grp.thread_rank() + row_start; i < row_end;
+             i += subwarp_grp.size()) {
+            int col = a.col_idxs[i];
+            ValueType val = a.values[i];
+
+            temp += val * b.values[col * b.stride + rhs];
         }
-        for (auto k = a.row_ptrs[row]; k < a.row_ptrs[row + 1]; ++k) {
-            auto val = a.values[k];
-            auto col = a.col_idxs[k];
-            for (int j = 0; j < b.num_rhs; ++j) {
-                c.values[row * c.stride + j] +=
-                    val * b.values[col * b.stride + j];
-            }
+
+        // subwarp_grp level reduction
+        for (int i = subwarp_grp.size() / 2; i > 0; i /= 2) {
```

Comment:
Yes, but I have changed row_end - row_start to a number rounded off to closest power of 2, to avoid missing few elements during reduction, when i is getting halved every iteration. Would that be okay?

Reply:
Thanks for going over it. I think it's okay, but there are now a lot of special functions there. I think it's fine for now though.